include/pstore/core/hamt_map_types.hpp

```diff
@@ -175,73 +205,56 @@ namespace pstore {
 
-                typed_address<internal_node> untag_internal_address () const noexcept {
-                    return typed_address<internal_node>::make (addr.absolute () &
-                                                               ~internal_node_bit);
+                address to_address () const noexcept {
+                    assert (is_address ());
+                    return addr_;
                 }
-                typed_address<linear_node> untag_linear_address () const noexcept {
-                    return typed_address<linear_node>::make (addr.absolute () & ~internal_node_bit);
-                }
-
-                address addr;
-                internal_node * internal;
-                linear_node * linear;
 
-            private:
-                static std::uintptr_t tag (void * const p) noexcept {
-                    return reinterpret_cast<std::uintptr_t> (p) | internal_node_bit | heap_node_bit;
+                template <typename T, typename = typename std::enable_if_t<
+                                          is_any_of<T, internal_node, linear_node>::value>>
+                typed_address<T> untag_address () const noexcept {
+                    return typed_address<T>::make (to_address ().absolute () & ~internal_node_bit);
                 }
-                std::uintptr_t tag () const noexcept { return tag (internal); }
 
-                static internal_node * tag_node (internal_node * const p) noexcept {
-                    return reinterpret_cast<internal_node *> (tag (p));
-                }
-                static linear_node * tag_node (linear_node * const p) noexcept {
-                    return reinterpret_cast<linear_node *> (tag (p));
+                template <typename Ptr, typename = typename std::enable_if_t<is_any_of<
+                                            typename std::pointer_traits<Ptr>::element_type,
+                                            internal_node, linear_node>::value>>
+                Ptr untag () const noexcept {
+                    return reinterpret_cast<Ptr> (reinterpret_cast<std::uintptr_t> (internal_) &
+                                                  ~internal_node_bit & ~heap_node_bit);
                 }
 
-                std::uintptr_t untag () const noexcept {
-                    return reinterpret_cast<std::uintptr_t> (internal) & ~internal_node_bit &
-                           ~heap_node_bit;
-                }
+            private:
+                template <typename Ptr, typename = typename std::enable_if_t<is_any_of<
+                                            typename std::pointer_traits<Ptr>::element_type,
+                                            internal_node, linear_node>::value>>
+                static Ptr tag (Ptr t) noexcept {
+                    return reinterpret_cast<Ptr> (reinterpret_cast<std::uintptr_t> (t) |
+                                                  internal_node_bit | heap_node_bit);
```

Comment:
Why tag both  `internal_node_bit` and `heap_node_bit`? Should the code tag `internal_node_bit` if the type is internal_node? tag the `heap_node_bit` if the type is heap_node?

Reply:
That’s what the original code did.